src/gdal/ogr_file_format.cpp

```diff
@@ -1408,5 +1416,354 @@ LatLon OgrFileImport::calcAverageLatLon(OGRDataSourceH data_source)
+			auto extension = extensions.mid(start, pos - start);
+			if (file_extn == QString::fromLatin1(extension))
+			{
+				auto cap_ogr = GDALGetMetadataItem(driver_data, GDAL_DCAP_VECTOR, nullptr);
+				if (qstrcmp(cap_ogr, "YES") != 0)
+					continue;
+				auto cap_open = GDALGetMetadataItem(driver_data, GDAL_DCAP_CREATE, nullptr);
+				if (qstrcmp(cap_open, "YES") != 0)
+					continue;
+				po_driver = driver_data;
+				break;
+			}
+		}
+	}
+#else
+	const char *psz_driver_name;
+	if (file_extn.compare(QString::fromLatin1("gpx"), Qt::CaseInsensitive))
+		psz_driver_name = "GPX";
+	else if (file_extn.compare(QString::fromLatin1("kml"), Qt::CaseInsensitive))
+		psz_driver_name = "KML";
+	else
+		psz_driver_name = "ESRI Shapefile";
+
+	po_driver = GDALGetDriverByName(psz_driver_name);
+#endif
+
+	if (!po_driver)
+		throw FileFormatException(tr("Couldn't find a driver for file extension %1").arg(file_extn));
+
+	if (qstrcmp(GDALGetDriverShortName(po_driver), "GPX") == 0)
+		need_wgs1984 = true;
+
+	// Check if map is georeferenced
+	const auto& georef = map->getGeoreferencing();
+	bool local_only = false;
+	if (georef.getState() == Georeferencing::Local)
+	{
+		local_only = true;
+		addWarning(tr("The map is not georeferenced.  Local georeferencing only."));
+	}
+
+	// Make sure GDAL can work with the georeferencing info
+	map_srs = ogr::unique_srs { OSRNewSpatialReference(nullptr) };
+	if (!local_only)
+	{
+		OSRSetProjCS(map_srs.get(), "Projected map SRS");
+		OSRSetWellKnownGeogCS(map_srs.get(), "WGS84");
+		auto spec = QByteArray(georef.getProjectedCRSSpec().toLatin1() + " +wktext");
+		if (OSRImportFromProj4(map_srs.get(), spec) != OGRERR_NONE)
+		{
+			local_only = true;
+			addWarning(tr("Failed to properly create the export georeferencing info.  Local georeferencing only."));
+		}
+	}
+
+	/**
+	 * Only certain drivers work without georeferencing info
+	 * Based on http://www.gdal.org/ogr_formats.html as of March 5, 2018
+	 * */
+	static const std::vector<QByteArray> local_drivers = {
+	    "ARCGEN", "BNA", "DWG", "DXF",
+	    "DXF", "CSV", "Geomedia", "INGRES",
+	    "OpenJUMP .jml", "DGN", "DGNv8", "ODS",
+	    "REC", "SEGY", "XLS", "XLSX"
+	};
+	if (local_only)
+	{
+		QByteArray driver_name = GDALGetDriverLongName(po_driver);
+		if (std::find(local_drivers.begin(), local_drivers.end(), driver_name) == local_drivers.end())
+		throw FileFormatException(tr("The %1 driver requires valid georefencing info.")
+		                          .arg(QString::fromLatin1(driver_name)));
+	}
+
+	// Create output dataset
+	po_ds = ogr::unique_datasource(OGR_Dr_CreateDataSource(po_driver, output_file.toLatin1(), nullptr));
+	if (!po_ds)
+		throw FileFormatException(tr("Failed to create dataset.  Error: %1").arg(QString::fromLatin1(CPLGetLastErrorMsg())));
+
+	/**
+	 * Create collections of point, line, and areas that are selected
+	 * These have to be separate as some drivers required them to be added in a specific order
```

Comment:
Creating collection means allocation memory. Even if it is not much, I would prefer to avoid it, and rather loop over the all objects multiple times, filtering as needed. I would build upon `Map::applyOnMatchingObjects()` (which takes two arguments, an "operation" function and a "condition" function). But I understand that this approach maybe a little bit hard when you have little experience with function objects and/or lambdas.

Maybe be it is also a disadvantage to always create layers in this way. For shape files, this means individual sets of files, so we would actually need to select a *directory* for exporting shape files, not a regular *file*.

Reply:
Ok, sounds good with regards to `Map::applyOnMatchingObjects()`  I guess I'll learn about lambdas and function objects :)

Most drivers don't support multiple geometry types in the same layer; adding the points, then the lines, then the area symbols was compatible with all drivers that I tested.  Do you have a preference on how symbols should be arranged into layers?  ie all of the same geometries in a singular layer like I have now, one layer per symbol, or something different?

With regards to shapefiles, should the layers be named <basename>-line, <basename>-point, etc so that we hopefully don't get conflicting names for files?